File: src/core/gradleManifest.ts
import { dependencyEntry } from './dependency';
import type { DependencyEntry, SourceRange } from './model';

type TokenKind = 'identifier' | 'string' | 'symbol' | 'newline';

interface Token {
  kind: TokenKind;
  value: string;
  start: number;
  end: number;
  /** Source range inside the quotes, for string tokens. */
  contentStart?: number;
}

interface ParsedDependency {
  noteKey: string;
  primaryRange: SourceRange;
  sourceRanges: SourceRange[];
}

const WRAPPERS = new Set(['platform', 'enforcedPlatform', 'testFixtures']);
const NON_LIBRARY_CATALOGS = new Set(['bundles', 'plugins', 'versions']);
const NON_CONFIGURATIONS = new Set(['val', 'var', 'def', 'if', 'for', 'while', 'when', 'return', 'println']);
const NON_MODULE_FACTORIES = new Set(['project', 'files', 'fileTree', 'gradleApi', 'localGroovy']);

/**
 * A deliberately small Groovy/Kotlin lexer. Gradle build scripts are programs,
 * so Pacmon only reads statically visible declarations and never evaluates the
 * script. Keeping comments, strings and balanced delimiters distinct avoids the
 * false positives a regular-expression parser would produce.
 */
function tokenize(text: string): Token[] {
  const tokens: Token[] = [];
  let i = 0;
  while (i < text.length) {
    const start = i;
    const ch = text[i]!;
    if (ch === '\r' || ch === '\n') {
      if (ch === '\r' && text[i + 1] === '\n') i++;
      i++;
      tokens.push({ kind: 'newline', value: '\n', start, end: i });
      continue;
    }
    if (/\s/.test(ch)) {
      i++;
      continue;
    }
    if (ch === '/' && text[i + 1] === '/') {
      i += 2;
      while (i < text.length && text[i] !== '\r' && text[i] !== '\n') i++;
      continue;
    }
    if (ch === '/' && text[i + 1] === '*') {
      i += 2;
      while (i < text.length && !(text[i] === '*' && text[i + 1] === '/')) {
        if (text[i] === '\r' || text[i] === '\n') {
          const lineStart = i;
          if (text[i] === '\r' && text[i + 1] === '\n') i++;
          i++;
          tokens.push({ kind: 'newline', value: '\n', start: lineStart, end: i });
        } else {
          i++;
        }
      }
      if (i < text.length) i += 2;
      continue;
    }
    if (ch === '"' || ch === "'") {
      const triple = text.slice(i, i + 3) === ch.repeat(3);
      const width = triple ? 3 : 1;
      i += width;
      const contentStart = i;
      while (i < text.length) {
        if (triple && text.slice(i, i + 3) === ch.repeat(3)) break;
        if (!triple && text[i] === ch) break;
        if (!triple && text[i] === '\\' && i + 1 < text.length) i += 2;
        else i++;
      }
      const contentEnd = i;
      if (i < text.length) i += width;
      tokens.push({
        kind: 'string',
        value: text.slice(contentStart, contentEnd),
        start,
        end: i,
        contentStart,
      });
      continue;
    }
    if (/[A-Za-z_$]/.test(ch)) {
      i++;
      while (i < text.length && /[A-Za-z0-9_$]/.test(text[i]!)) i++;
      tokens.push({ kind: 'identifier', value: text.slice(start, i), start, end: i });
      continue;
    }
    i++;
    tokens.push({ kind: 'symbol', value: ch, start, end: i });
  }
  return tokens;
}

function nextCode(tokens: readonly Token[], index: number, end = tokens.length): number {
  while (index < end && tokens[index]?.kind === 'newline') index++;
  return index;
}

function matching(tokens: readonly Token[], open: number, end: number, left: string, right: string): number | undefined {
  let depth = 0;
  for (let i = open; i < end; i++) {
    const token = tokens[i]!;
    if (token.kind !== 'symbol') continue;
    if (token.value === left) depth++;
    else if (token.value === right && --depth === 0) return i;
  }
  return undefined;
}

function decodeSimpleString(value: string): string {
  return value.replace(/\\([\\'"nrt])/g, (_match, escaped: string) => {
    if (escaped === 'n') return '\n';
    if (escaped === 'r') return '\r';
    if (escaped === 't') return '\t';
    return escaped;
  });
}

function stringRange(token: Token, start = 0, length = token.value.length): SourceRange {
  return { offset: (token.contentStart ?? token.start) + start, length };
}

function coordinateFrom(token: Token): ParsedDependency | undefined {
  if (token.kind !== 'string') return undefined;
  const first = token.value.indexOf(':');
  const second = token.value.indexOf(':', first + 1);
  const end = second === -1 ? token.value.length : second;
  if (first <= 0 || end <= first + 1) return undefined;
  const group = decodeSimpleString(token.value.slice(0, first));
  const artifact = decodeSimpleString(token.value.slice(first + 1, end));
  if (!group || !artifact) return undefined;
  const groupRange = stringRange(token, 0, first);
  const artifactRange = stringRange(token, first + 1, end - first - 1);
  return {
    noteKey: `${group}:${artifact}`,
    primaryRange: artifactRange,
    sourceRanges: [groupRange, artifactRange],
  };
}

function mapNotation(tokens: readonly Token[], start: number, end: number): ParsedDependency | undefined {
  let group: Token | undefined;
  let name: Token | undefined;
  for (let i = start; i < end; i++) {
    const key = tokens[i];
    if (key?.kind !== 'identifier' || (key.value !== 'group' && key.value !== 'name')) continue;
    const separator = nextCode(tokens, i + 1, end);
    if (![':', '='].includes(tokens[separator]?.value ?? '')) continue;
    const valueIndex = nextCode(tokens, separator + 1, end);
    const value = tokens[valueIndex];
    if (value?.kind !== 'string') continue;
    if (key.value === 'group') group = value;
    else name = value;
  }
  if (!group || !name) return undefined;
  const groupName = decodeSimpleString(group.value);
  const artifactName = decodeSimpleString(name.value);
  if (!groupName || !artifactName) return undefined;
  const groupRange = stringRange(group);
  const nameRange = stringRange(name);
  return {
    noteKey: `${groupName}:${artifactName}`,
    primaryRange: nameRange,
    sourceRanges: [groupRange, nameRange],
  };
}

function catalogAlias(tokens: readonly Token[], start: number, end: number): ParsedDependency | undefined {
  for (let i = start; i < end; i++) {
    const root = tokens[i];
    if (root?.kind !== 'identifier' || root.value !== 'libs') continue;
    const parts = ['libs'];
    let last = root;
    let cursor = i + 1;
    while (cursor + 1 < end && tokens[cursor]?.value === '.' && tokens[cursor + 1]?.kind === 'identifier') {
      const part = tokens[cursor + 1]!;
      if (part.value === 'get' && tokens[nextCode(tokens, cursor + 2, end)]?.value === '(') break;
      parts.push(part.value);
      last = part;
      cursor += 2;
    }
    if (parts.length < 2 || NON_LIBRARY_CATALOGS.has(parts[1]!)) continue;
    const range = { offset: root.start, length: last.end - root.start };
    return { noteKey: parts.join('.'), primaryRange: range, sourceRanges: [range] };
  }
  return undefined;
}

function expression(tokens: readonly Token[], start: number, end: number): ParsedDependency | undefined {
  for (let i = start; i < end; i++) {
    const token = tokens[i];
    if (token?.kind === 'identifier' && NON_MODULE_FACTORIES.has(token.value)) return undefined;
  }
  const mapped = mapNotation(tokens, start, end);
  if (mapped) return mapped;
  const alias = catalogAlias(tokens, start, end);
  if (alias) return alias;
  for (let i = start; i < end; i++) {
    const token = tokens[i]!;
    if (token.kind === 'string') {
      const coordinate = coordinateFrom(token);
      if (coordinate) return coordinate;
    }
    if (token.kind === 'identifier' && WRAPPERS.has(token.value)) {
      const open = nextCode(tokens, i + 1, end);
      if (tokens[open]?.value === '(') {
        const close = matching(tokens, open, end, '(', ')');
        if (close !== undefined) {
          const nested = expression(tokens, open + 1, close);
          if (nested) return nested;
        }
      }
    }
  }
  return undefined;
}

function topLevelComma(tokens: readonly Token[], start: number, end: number): number | undefined {
  let parens = 0;
  let brackets = 0;
  for (let i = start; i < end; i++) {
    const value = tokens[i]?.value;
    if (value === '(') parens++;
    else if (value === ')') parens--;
    else if (value === '[') brackets++;
    else if (value === ']') brackets--;
    else if (value === ',' && parens === 0 && brackets === 0) return i;
  }
  return undefined;
}

function dependencyBody(tokens: readonly Token[], start: number, end: number): DependencyEntry[] {
  const out: DependencyEntry[] = [];
  let i = start;
  let braceDepth = 0;
  while (i < end) {
    const token = tokens[i]!;
    if (token.value === '{') {
      braceDepth++;
      i++;
      continue;
    }
    if (token.value === '}') {
      braceDepth--;
      i++;
      continue;
    }
    if (braceDepth !== 0 || (token.kind !== 'identifier' && token.kind !== 'string')) {
      i++;
      continue;
    }
    const configuration = token;
    if (configuration.kind === 'identifier' && NON_CONFIGURATIONS.has(configuration.value)) {
      while (i < end && tokens[i]?.kind !== 'newline' && tokens[i]?.value !== ';') i++;
      i++;
      continue;
    }
    const afterName = nextCode(tokens, i + 1, end);
    if (tokens[afterName]?.value === '{') {
      i++;
      continue;
    }

    let expressionStart = afterName;
    let expressionEnd = afterName;
    let statementEnd = afterName;
    if (tokens[afterName]?.value === '(') {
      const close = matching(tokens, afterName, end, '(', ')');
      if (close === undefined) {
        i++;
        continue;
      }
      expressionStart = afterName + 1;
      expressionEnd = close;
      statementEnd = close + 1;
    } else {
      let parens = 0;
      let brackets = 0;
      let cursor = afterName;
      for (; cursor < end; cursor++) {
        const current = tokens[cursor]!;
        if (current.value === '(') parens++;
        else if (current.value === ')') parens--;
        else if (current.value === '[') brackets++;
        else if (current.value === ']') brackets--;
        if (parens === 0 && brackets === 0 && (current.kind === 'newline' || current.value === ';' || current.value === '{')) break;
      }
      expressionEnd = cursor;
      statementEnd = cursor;
    }

    let scope = configuration.kind === 'string' ? decodeSimpleString(configuration.value) : configuration.value;
    if (scope === 'add') {
      const comma = topLevelComma(tokens, expressionStart, expressionEnd);
      const scopeToken = tokens[nextCode(tokens, expressionStart, expressionEnd)];
      if (comma === undefined || scopeToken?.kind !== 'string') {
        i = Math.max(statementEnd, i + 1);
        continue;
      }
      scope = decodeSimpleString(scopeToken.value);
      expressionStart = comma + 1;
    }

    const parsed = expression(tokens, expressionStart, expressionEnd);
    if (parsed && scope) {
      out.push(dependencyEntry(
        parsed.noteKey,
        scope,
        parsed.primaryRange,
        parsed.sourceRanges,
        parsed.noteKey,
        { offset: configuration.start, length: configuration.end - configuration.start },
      ));
    }
    i = Math.max(statementEnd, i + 1);
  }
  return out;
}

export function extractGradleDependencies(text: string): DependencyEntry[] {
  const tokens = tokenize(text);
  const out: DependencyEntry[] = [];
  const blocks: string[] = [];
  for (let i = 0; i < tokens.length; i++) {
    const token = tokens[i]!;
    if (token.value === '}') {
      blocks.pop();
      continue;
    }
    if (token.value === '{') {
      const previous = tokens[i - 1];
      blocks.push(previous?.kind === 'identifier' ? previous.value : '');
      continue;
    }
    if (token.kind !== 'identifier' || token.value !== 'dependencies' || blocks.includes('buildscript')) continue;
    const open = nextCode(tokens, i + 1);
    if (tokens[open]?.value !== '{') continue;
    const close = matching(tokens, open, tokens.length, '{', '}');
    if (close === undefined) continue;
    out.push(...dependencyBody(tokens, open + 1, close));
    i = close;
  }
  return out;
}

File: jetbrains/src/main/kotlin/dev/pacmon/jetbrains/core/GradleManifest.kt
package dev.pacmon.jetbrains.core

private enum class GradleTokenKind { IDENTIFIER, STRING, SYMBOL, NEWLINE }

private data class GradleToken(
    val kind: GradleTokenKind,
    val value: String,
    val start: Int,
    val end: Int,
    val contentStart: Int? = null,
)

private data class ParsedGradleDependency(
    val noteKey: String,
    val primaryRange: SourceRange,
    val sourceRanges: List<SourceRange>,
)

/**
 * Static Groovy/Kotlin DSL reader. Gradle scripts are executable programs, so
 * Pacmon deliberately reads only source-visible dependency declarations and
 * never evaluates a script or invokes Gradle.
 */
object GradleManifestAdapter : ManifestAdapter {
    override val kind = ManifestKind.GRADLE
    override val fileNames = listOf("build.gradle.kts", "build.gradle")
    override val notesRelativePath = ".pacmon/gradle/DEPENDENCY-NOTES.md"

    override fun normalizeNoteKey(raw: String): String = ManifestRegistry.stripName(raw)

    override fun extractDependencies(text: String): List<DependencyEntry> {
        val tokens = tokenizeGradle(text)
        val result = mutableListOf<DependencyEntry>()
        val blocks = mutableListOf<String>()
        var index = 0
        while (index < tokens.size) {
            val token = tokens[index]
            when (token.value) {
                "}" -> {
                    if (blocks.isNotEmpty()) blocks.removeAt(blocks.lastIndex)
                    index++
                    continue
                }
                "{" -> {
                    val previous = tokens.getOrNull(index - 1)
                    blocks.add(if (previous?.kind == GradleTokenKind.IDENTIFIER) previous.value else "")
                    index++
                    continue
                }
            }
            if (token.kind != GradleTokenKind.IDENTIFIER || token.value != "dependencies" || "buildscript" in blocks) {
                index++
                continue
            }
            val open = nextGradleCode(tokens, index + 1)
            if (tokens.getOrNull(open)?.value != "{") {
                index++
                continue
            }
            val close = matchingGradle(tokens, open, tokens.size, "{", "}")
            if (close == null) {
                index++
                continue
            }
            result.addAll(parseDependencyBody(tokens, open + 1, close))
            index = close + 1
        }
        return result
    }
}

private val gradleWrappers = setOf("platform", "enforcedPlatform", "testFixtures")
private val nonLibraryCatalogs = setOf("bundles", "plugins", "versions")
private val nonConfigurations = setOf("val", "var", "def", "if", "for", "while", "when", "return", "println")
private val nonModuleFactories = setOf("project", "files", "fileTree", "gradleApi", "localGroovy")

private fun tokenizeGradle(text: String): List<GradleToken> {
    val tokens = mutableListOf<GradleToken>()
    var index = 0
    while (index < text.length) {
        val start = index
        val char = text[index]
        if (char == '\r' || char == '\n') {
            if (char == '\r' && text.getOrNull(index + 1) == '\n') index++
            index++
            tokens.add(GradleToken(GradleTokenKind.NEWLINE, "\n", start, index))
            continue
        }
        if (char.isWhitespace()) {
            index++
            continue
        }
        if (char == '/' && text.getOrNull(index + 1) == '/') {
            index += 2
            while (index < text.length && text[index] != '\r' && text[index] != '\n') index++
            continue
        }
        if (char == '/' && text.getOrNull(index + 1) == '*') {
            index += 2
            while (index < text.length && !(text[index] == '*' && text.getOrNull(index + 1) == '/')) {
                if (text[index] == '\r' || text[index] == '\n') {
                    val lineStart = index
                    if (text[index] == '\r' && text.getOrNull(index + 1) == '\n') index++
                    index++
                    tokens.add(GradleToken(GradleTokenKind.NEWLINE, "\n", lineStart, index))
                } else {
                    index++
                }
            }
            if (index < text.length) index += 2
            continue
        }
        if (char == '"' || char == '\'') {
            val triple = text.regionMatches(index, char.toString().repeat(3), 0, 3)
            val width = if (triple) 3 else 1
            index += width
            val contentStart = index
            while (index < text.length) {
                if (triple && text.regionMatches(index, char.toString().repeat(3), 0, 3)) break
                if (!triple && text[index] == char) break
                if (!triple && text[index] == '\\' && index + 1 < text.length) index += 2 else index++
            }
            val contentEnd = index
            if (index < text.length) index += width
            tokens.add(
                GradleToken(
                    GradleTokenKind.STRING,
                    text.substring(contentStart, contentEnd),
                    start,
                    index,
                    contentStart,
                ),
            )
            continue
        }
        if (char.isLetter() || char == '_' || char == '$') {
            index++
            while (index < text.length && (text[index].isLetterOrDigit() || text[index] == '_' || text[index] == '$')) index++
            tokens.add(GradleToken(GradleTokenKind.IDENTIFIER, text.substring(start, index), start, index))
            continue
        }
        index++
        tokens.add(GradleToken(GradleTokenKind.SYMBOL, char.toString(), start, index))
    }
    return tokens
}

private fun nextGradleCode(tokens: List<GradleToken>, from: Int, end: Int = tokens.size): Int {
    var index = from
    while (index < end && tokens[index].kind == GradleTokenKind.NEWLINE) index++
    return index
}

private fun matchingGradle(
    tokens: List<GradleToken>,
    open: Int,
    end: Int,
    left: String,
    right: String,
): Int? {
    var depth = 0
    for (index in open until end) {
        val token = tokens[index]
        if (token.kind != GradleTokenKind.SYMBOL) continue
        if (token.value == left) depth++
        else if (token.value == right) {
            depth--
            if (depth == 0) return index
        }
    }
    return null
}

private fun decodeGradleString(value: String): String {
    val result = StringBuilder()
    var index = 0
    while (index < value.length) {
        if (value[index] != '\\' || index + 1 >= value.length) {
            result.append(value[index++])
            continue
        }
        val escaped = value[index + 1]
        result.append(
            when (escaped) {
                'n' -> '\n'
                'r' -> '\r'
                't' -> '\t'
                else -> escaped
            },
        )
        index += 2
    }
    return result.toString()
}

private fun stringRange(token: GradleToken, start: Int = 0, length: Int = token.value.length) =
    SourceRange((token.contentStart ?: token.start) + start, length)

private fun coordinateFrom(token: GradleToken): ParsedGradleDependency? {
    if (token.kind != GradleTokenKind.STRING) return null
    val first = token.value.indexOf(':')
    val second = token.value.indexOf(':', first + 1)
    val end = if (second == -1) token.value.length else second
    if (first <= 0 || end <= first + 1) return null
    val group = decodeGradleString(token.value.substring(0, first))
    val artifact = decodeGradleString(token.value.substring(first + 1, end))
    if (group.isEmpty() || artifact.isEmpty()) return null
    val groupRange = stringRange(token, 0, first)
    val artifactRange = stringRange(token, first + 1, end - first - 1)
    return ParsedGradleDependency("$group:$artifact", artifactRange, listOf(groupRange, artifactRange))
}

private fun mapNotation(tokens: List<GradleToken>, start: Int, end: Int): ParsedGradleDependency? {
    var group: GradleToken? = null
    var name: GradleToken? = null
    for (index in start until end) {
        val key = tokens[index]
        if (key.kind != GradleTokenKind.IDENTIFIER || key.value !in setOf("group", "name")) continue
        val separator = nextGradleCode(tokens, index + 1, end)
        if (tokens.getOrNull(separator)?.value !in setOf(":", "=")) continue
        val valueIndex = nextGradleCode(tokens, separator + 1, end)
        val value = tokens.getOrNull(valueIndex)
        if (value?.kind != GradleTokenKind.STRING) continue
        if (key.value == "group") group = value else name = value
    }
    val groupToken = group ?: return null
    val nameToken = name ?: return null
    val groupName = decodeGradleString(groupToken.value)
    val artifactName = decodeGradleString(nameToken.value)
    if (groupName.isEmpty() || artifactName.isEmpty()) return null
    val groupRange = stringRange(groupToken)
    val nameRange = stringRange(nameToken)
    return ParsedGradleDependency("$groupName:$artifactName", nameRange, listOf(groupRange, nameRange))
}

private fun catalogAlias(tokens: List<GradleToken>, start: Int, end: Int): ParsedGradleDependency? {
    for (index in start until end) {
        val root = tokens[index]
        if (root.kind != GradleTokenKind.IDENTIFIER || root.value != "libs") continue
        val parts = mutableListOf("libs")
        var last = root
        var cursor = index + 1
        while (cursor + 1 < end && tokens[cursor].value == "." && tokens[cursor + 1].kind == GradleTokenKind.IDENTIFIER) {
            val part = tokens[cursor + 1]
            if (part.value == "get" && tokens.getOrNull(nextGradleCode(tokens, cursor + 2, end))?.value == "(") break
            parts.add(part.value)
            last = part
            cursor += 2
        }
        if (parts.size < 2 || parts[1] in nonLibraryCatalogs) continue
        val range = SourceRange(root.start, last.end - root.start)
        return ParsedGradleDependency(parts.joinToString("."), range, listOf(range))
    }
    return null
}

private fun parseExpression(tokens: List<GradleToken>, start: Int, end: Int): ParsedGradleDependency? {
    if ((start until end).any { tokens[it].kind == GradleTokenKind.IDENTIFIER && tokens[it].value in nonModuleFactories }) {
        return null
    }
    mapNotation(tokens, start, end)?.let { return it }
    catalogAlias(tokens, start, end)?.let { return it }
    for (index in start until end) {
        val token = tokens[index]
        if (token.kind == GradleTokenKind.STRING) coordinateFrom(token)?.let { return it }
        if (token.kind == GradleTokenKind.IDENTIFIER && token.value in gradleWrappers) {
            val open = nextGradleCode(tokens, index + 1, end)
            if (tokens.getOrNull(open)?.value == "(") {
                val close = matchingGradle(tokens, open, end, "(", ")")
                if (close != null) parseExpression(tokens, open + 1, close)?.let { return it }
            }
        }
    }
    return null
}

private fun topLevelComma(tokens: List<GradleToken>, start: Int, end: Int): Int? {
    var parens = 0
    var brackets = 0
    for (index in start until end) {
        when (tokens[index].value) {
            "(" -> parens++
            ")" -> parens--
            "[" -> brackets++
            "]" -> brackets--
            "," -> if (parens == 0 && brackets == 0) return index
        }
    }
    return null
}

private fun parseDependencyBody(tokens: List<GradleToken>, start: Int, end: Int): List<DependencyEntry> {
    val result = mutableListOf<DependencyEntry>()
    var index = start
    var braceDepth = 0
    while (index < end) {
        val token = tokens[index]
        if (token.value == "{") {
            braceDepth++
            index++
            continue
        }
        if (token.value == "}") {
            braceDepth--
            index++
            continue
        }
        if (braceDepth != 0 || token.kind !in setOf(GradleTokenKind.IDENTIFIER, GradleTokenKind.STRING)) {
            index++
            continue
        }
        val configuration = token
        if (configuration.kind == GradleTokenKind.IDENTIFIER && configuration.value in nonConfigurations) {
            while (index < end && tokens[index].kind != GradleTokenKind.NEWLINE && tokens[index].value != ";") index++
            index++
            continue
        }
        val afterName = nextGradleCode(tokens, index + 1, end)
        if (tokens.getOrNull(afterName)?.value == "{") {
            index++
            continue
        }

        var expressionStart = afterName
        var expressionEnd: Int
        val statementEnd: Int
        if (tokens.getOrNull(afterName)?.value == "(") {
            val close = matchingGradle(tokens, afterName, end, "(", ")")
            if (close == null) {
                index++
                continue
            }
            expressionStart = afterName + 1
            expressionEnd = close
            statementEnd = close + 1
        } else {
            var parens = 0
            var brackets = 0
            var cursor = afterName
            while (cursor < end) {
                val current = tokens[cursor]
                when (current.value) {
                    "(" -> parens++
                    ")" -> parens--
                    "[" -> brackets++
                    "]" -> brackets--
                }
                if (parens == 0 && brackets == 0 &&
                    (current.kind == GradleTokenKind.NEWLINE || current.value in setOf(";", "{"))
                ) break
                cursor++
            }
            expressionEnd = cursor
            statementEnd = cursor
        }

        var scope = if (configuration.kind == GradleTokenKind.STRING) {
            decodeGradleString(configuration.value)
        } else {
            configuration.value
        }
        if (scope == "add") {
            val comma = topLevelComma(tokens, expressionStart, expressionEnd)
            val scopeToken = tokens.getOrNull(nextGradleCode(tokens, expressionStart, expressionEnd))
            if (comma == null || scopeToken?.kind != GradleTokenKind.STRING) {
                index = maxOf(statementEnd, index + 1)
                continue
            }
            scope = decodeGradleString(scopeToken.value)
            expressionStart = comma + 1
        }

        parseExpression(tokens, expressionStart, expressionEnd)?.let { parsed ->
            if (scope.isNotEmpty()) {
                result.add(
                    dependencyEntry(
                        parsed.noteKey,
                        scope,
                        parsed.primaryRange,
                        parsed.sourceRanges,
                        parsed.noteKey,
                        SourceRange(configuration.start, configuration.end - configuration.start),
                    ),
                )
            }
        }
        index = maxOf(statementEnd, index + 1)
    }
    return result
}
